### PR TITLE
feat: added util guard Rate Limit commands

### PR DESCRIPTION
**Pull request by VictoriqueMoe**
This PR adds a guard that can be applied to commands to rate limit them.

## utilities

This will add the ability to rate limit commands

Added:
Example
Docs
The Guard
TimedSet util class that is a set with a timer on each element

### CODE REVIEW

File: packages/utilities/examples/guards/discords/RateLimit.ts [hunk 1]
```diff
@@ -2,25 +2,48 @@ import type { CommandInteraction } from "discord.js";
 import type { SimpleCommandMessage } from "discordx";
 import { Discord, Guard, SimpleCommand, Slash } from "discordx";
 
-import { RateLimit, TIME_UNIT } from "../../../src/index.js";
```
Comment (oceanroleplay) [inline packages/utilities/examples/guards/discords/RateLimit.ts:5]: revert

File: packages/utilities/src/guards/Rate Limiter/RateLimit.ts [hunk 2]
```diff
@@ -55,39 +79,36 @@ export function RateLimit(
...
     const arr = _timer.rawSet;
     return arr.find(
-      (v: TimeOutEntry) => v.userId === userId && v.guildId === guildId
+        (v: TimeOutEntry) => v.userId === userId && v.guildId === guildId
     );
-  };
+  }
 
-  const post = async (
-    arg: BaseCommandInteraction | SimpleCommandMessage,
```
Comment (oceanroleplay) [inline packages/utilities/src/guards/Rate Limiter/RateLimit.ts:81]: rever

File: packages/utilities/src/guards/Rate Limiter/RateLimit.ts [hunk 3]
```diff
@@ -100,19 +120,20 @@ export function RateLimit(
...
+      _timer.refresh(fromArray);
     } else {
-      fromArray = new TimeOutEntry(memberId, guildId);
+      fromArray = new TimeOutEntry(memberId, guildId, rateValue);
       _timer.add(fromArray);
     }
-
     return next();
   };
-}
```
Comment (oceanroleplay) [inline packages/utilities/src/guards/Rate Limiter/RateLimit.ts:118]: revert

File: packages/utilities/src/guards/Rate Limiter/RateLimit.ts [hunk 4]
```diff
@@ -61,38 +84,37 @@ export function RateLimit(
...
+  ): Promise<void> {
     if (arg instanceof SimpleCommandMessage) {
       await arg?.message.reply(msg);
     } else {
-      await replyOrFollowUp(arg, msg);
-      return;
+      return replyOrFollowUp(arg, msg);
     }
-  };
+
```
Comment (oceanroleplay) [inline packages/utilities/src/guards/Rate Limiter/RateLimit.ts:112]: no newline

File: packages/utilities/src/guards/Rate Limiter/RateLimit.ts [hunk 5]
```diff
@@ -61,38 +84,37 @@ export function RateLimit(
...
+  }
```
Comment (oceanroleplay) [inline packages/utilities/src/guards/Rate Limiter/RateLimit.ts:114]: revert

File: packages/utilities/src/guards/Rate Limiter/RateLimit.ts [hunk 6]
```diff
@@ -61,38 +84,37 @@ export function RateLimit(
...
 
   return (arg, client, next) => {
     let memberId: string | null | undefined;
     let guildId: string | null;
-
```
Comment (oceanroleplay) [inline packages/utilities/src/guards/Rate Limiter/RateLimit.ts:95]: revert

File: packages/utilities/src/guards/Rate Limiter/RateLimit.ts (hunk 3 above)
Comment (VictoriqueMoe) [reply packages/utilities/src/guards/Rate Limiter/RateLimit.ts:118]: why? there should be a semicolon here

Comment (oceanroleplay) [reply packages/utilities/src/guards/Rate Limiter/RateLimit.ts:118]: why?

Comment (oceanroleplay) [reply packages/utilities/src/guards/Rate Limiter/RateLimit.ts:118]: oh you have used arrow function here, so use normal function instead.

**Pull request merged**